3단계 - 로또(자동) 리뷰 요청드립니다

tlaqk229 opened the pull request:
안녕하세요!
지난번 2단계 미션에 주신 리뷰 코멘트 반영 후 3단계 미션진행하였습니다
리뷰 부탁드립니다
감사합니다~!

tlaqk229 committed:
refactor(StringCalculator): step2 리뷰 코멘트 반영

- 메소드 이름 수정 (의미 불분명한 메소드들)
- 커스텀 구분자를 찾지 못한 경우 기본 구분자로 분리하도록 수정
- 런타임 에러 메세지 추가
- depth 2단계로 작성된 부분 depth 1이 되도록 수정

tlaqk229 committed:
docs(mission3): 3단계 미션 요구사항 & 기능목록 작성

tlaqk229 committed:
docs(mission3): 기능목록 수정

tlaqk229 committed:
test(lotto): 도메인 및 vo 단위테스트 작성

tlaqk229 committed:
feat(view,dto): 사용자 입력 구현 및 관련 dto 추가

tlaqk229 committed:
feat(view,dto,vo): 출력 클래스 추가 및 관련 dto,vo 추가/수정

tlaqk229 committed:
feat(controller,service): 로또구입/당첨확인 실행 controller,service 추가

tlaqk229 committed:
feat(service): 로또 구매 개수 계산

tlaqk229 committed:
feat(service): 구매 개수 만큼 로또 발급

tlaqk229 committed:
feat(vo): 로또 당첨 결과관련 vo 추가

- WinCriterion : 당첨 기준 (등수, 일치개수, 상금)
- WinCriteria : WinCriterion 모음 (당첨 기준 전체)
- MatchCounts : 발급된 로또들 각각 당첨번호와 일치하는 번호 개수 모음
- WinResult : 당첨 기준별 해당 로또 개수
- Profit : 당첨 수익률

tlaqk229 committed:
feat(service): 로또 당첨 결과 확인, 수익률 계산

tlaqk229 committed:
test(model): 로또 비교 및 유효성 체크(중복/개수초과) 테스트 추가

tlaqk229 committed:
feat(view): 발급된 로또 및 당첨결과 출력

tlaqk229 committed:
feat(vo): 당첨로또 번호 개수 확인 로직 추가

- 로또 번호 개수를 부족하게 입력하는 경우 예외처리

tlaqk229 committed:
fix(controller): service 메소드 기능분리로 인한 controller 수정

tlaqk229 committed:
refactor(controller): 기능에 따라 메소드 분리

tlaqk229 committed:
refactor(service,vo): 중복되는 불필요한 생성자 제거, 대체

tlaqk229 committed:
refactor(all): 자바 코드 컨벤션 - code reformat, import 최적화

ohtaeg commented on README.md:
```diff
@@ -60,3 +60,52 @@
+
+---
+### 3단계 - 로또(자동)
+#### 기능 요구사항
```
전체적으로 validation을 잘해주셨어요! 👍 
(구입 금액, 로또 번호의 중복, 로또 번호 범위)

ohtaeg commented on src/main/java/lotto/model/domain/LottoRun.java:
```diff
@@ -0,0 +1,20 @@
+import lotto.model.vo.PurchaseCount;
+import lotto.model.vo.WinLotto;
+
+public class LottoRun {
```
해당 클래스에 대해 궁금한 부분이 있는데요 :)
- 이 클래스가 가지고 있는 책임 혹은 하려는 행위는 무엇인지 말씀해주실 수 있으신가요 😄 ?
- LottoRun만 도메인으로 분리된 이유가 궁금해요!
- 윤지님께서 생각하시는 도메인이란 무엇인지 궁금합니다 :D

tlaqk229 replied:
TDD 방식으로 코딩하려 하다보니, 기능목록에 작성된 로직을 테스트하는 코드를 작성하고 이에 대한 클래스(LottoRun)를 생성하게 되었습니다. 이후 프로그램을 작성해나가는 과정에서 핵심로직 구현이 VO의 독립성을 지키도록 작성하다 보니 다른 도메인 로직들은 VO 안에 구현되고, 최초에 테스트코드를 먼저 작성한 부분만 동떨어지게 LottoRun에 구현된 상태로 남게 되었습니다. 생각해보니 각 도메인 객체에 해당하는 로직들이라, LottoRun 클래스는 제거하고 관련된 도메인 객체(VO) 내에 각 로직을 담는 방향으로 수정하였습니다.

ohtaeg commented on src/main/java/lotto/model/domain/LottoRun.java:
```diff
@@ -0,0 +1,20 @@
+
+public class LottoRun {
+    public static PurchaseCount getPurchaseCount(long buyAmount) {
+        return new PurchaseCount(buyAmount / LottoConstants.LOTTO_UNIT_PRICE);
```
이 부분은 같이 고민해보면 좋을 것 같아요!
구입 금액인 PurchaseAmount에서 getter로 데이터를 외부로 노출하여 Count를 계산하는 것도 좋지만
PurchaseAmount가 갖고있는 데이터를 외부로 노출하지 말고 PurchaseAmount에서 카운트를 계산하는 것에 대해 어떻게 생각하시나요 :) ?

tlaqk229 replied:
PurchaseAmount는 DTO로 만든 객체이므로 그대로 유지하였고, DTO와 VO 분리를 위해 새로운 객체 PurchaseInfo를 만들었습니다. 그리고 PurchaseInfo 생성자에 금액을 전달하면 카운트를 계산하여 세팅하도록 구현해보았습니다.

ohtaeg commented on src/main/java/lotto/model/domain/LottoRun.java:
```diff
@@ -0,0 +1,20 @@
+
+    public static int countMatchNumber(WinLotto winLotto, Lotto userLotto) {
+        return winLotto.compareWithLotto(userLotto);
+    }
```
https://github.com/next-step/java-lotto-pro/pull/906#discussion_r1013585258
해당 피드백과 같이 고려해서 개인적으로는 코드가 중복되는 느낌이 드는데요..!
```
// WinLotto.class
public int compareWithLotto(Lotto userLotto) {
    return winLotto.compare(userLotto);
}
```
LottoRun을 통해 호출이 더 늘어나는 느낌이 드는데요!
위 피드백과 같이 고려하여 윤지님의 의견을 같이 공유해주시면 좋을 것 같아요 :)

tlaqk229 replied:
앞서 LottoRun만 도메인으로 분리된 이유를 물어주신 데에 대한 답과 같습니다.
기능목록에 따라 테스트코드 작성 -> 도메인 로직 구현 클래스 생성을 해둔 상태에서 프로그램을 작성해나가다 보니 이렇게 작성되었습니다.
LottoRun은 제거하였고 불필요한 호출이 이루어지는 부분은 직접 VO의 메소드를 호출하도록 수정하였습니다.

ohtaeg commented on src/main/java/lotto/model/vo/Lotto.java:
```diff
@@ -0,0 +1,95 @@
+package lotto.model.vo;
```
개인적으로는 VO 패키지에 있는 몇몇의 클래스들은 도메인이라고 생각이 들어 한번 같이 의논하면 좋을 것 같아 코멘트를 남겨봅니다 :)
제가 생각하는 도메인은 요구사항을 소프트웨어로 해결하고자 만든 모델링의 결과물이라고 생각하는데요!
로또라는 요구사항에 맞게 Lotto라는 모델링이 되었다고 생각해요 :)
이 클래스는 도메인 처리의 중심이 되는 로직 및 개념이라 생각하여 도메인이라고 생각하는데요!

VO 같은 경우는 좀 더 값이라는 개념에 더 가까우며 도메인들이 서로 대화하기 위해 주고받는 데이터들이라고 생각합니다 :)
그래서 도메인과 VO를 분리해보면 좋을 것 같은데 윤지님이 생각하시는 도메인과 vo가 무엇인지 알려주시면 좋을 것 같아요 !! 😄

tlaqk229 replied:
도메인은 로직구현된 부분, VO는 값을 담고 있는 객체라고 잘 못 이해하고 있었습니다. 여러 자료를 찾아보고 도메인이란 로직만을 의미하는 것이 아니라 소프트웨어로 담고자 하는 문제 영역 전체라는 것을 알게 되었고, 이를 구현하기 위한 로직, 데이터 모두 도메인에 해당한다는 것을 알게되었습니다. 즉, 제가 단순히 VO라고 생각했던 객체들도 문제영역과 관련된 값과 로직을 담고 있고 이 또한 도메인에 해당한다는 생각을 하게 되었습니다. 따라서 vo 패키지는 domain으로 명칭을 수정했으며, 기존 domain 패키지 안에 있던 LottoRun 클래스는 제거하고 안에 구현된 로직은 각 도메인 객체 안에 담았습니다.

ohtaeg commented on src/main/java/lotto/model/vo/Lotto.java:
```diff
@@ -0,0 +1,95 @@
+        }
+        System.out.print(sj);
+        System.out.println(LottoConstants.LOTTO_PRINT_END);
+    }
```
이 부분은 개인 차이가 있기 때문에 한번 같이 얘기해보면 좋을 것 같아요 :)
개인적으로 위와 같이 해당 print 내부에서 구현된 출력문은 View 영역에 좀 더 가깝다고 생각하는데요!

핵심 비지니스 로직을 가지는 도메인 객체에서 어떻게 그려지는지에 대한 관심은 분리되면 좋지 않을까 합니다 😄
도메인을 I/O View와 강한 결합을 가지도록 설계를 두면 안될것 같다는 생각인데
윤지님은 어떻게 생각하시는지 궁금하네요 :)

tlaqk229 replied:
getter, setter를 사용하지 않고 객체에 메세지를 보내 처리해야한다는 생각에서 비롯된 코드입니다. 프리코스 1주차 피드백 영상과 1주차 라이브 강의 영상을 다시 보다보니 getter, setter 사용을 지양하라는 것이지 getter를 무조건 사용하지 말라는 얘기는 아니라는 내용을 확인할 수 있었습니다. getter를 사용하지 않으려다 보니 vo 안에 print 로직을 구현해두었는데 코멘트 주신대로 이 부분은 view에 해당하므로 getter를 사용하여 ResultView에 출력 로직을 구현하도록 수정했습니다.

ohtaeg commented on src/main/java/lotto/model/vo/LottoNumber.java:
```diff
@@ -0,0 +1,64 @@
+import lotto.model.constants.ErrorMessage;
+import lotto.model.constants.LottoConstants;
+
+public class LottoNumber implements Comparable<LottoNumber> {
```
작은 단위로 객체를 잘 분리해주셨어요 👍 
Comparable까지 구현 👍

ohtaeg commented on src/main/java/lotto/model/vo/LottoNumber.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    public int compareTo(LottoNumber o) {
+        return this.lottoNumber - o.lottoNumber;
```
Integer.compare도 활용 해보시면 좋을 것 같아요 :)

tlaqk229 replied:
Integer도 compare 활용할 수 있다는 점을 놓쳤네요. 조언 감사합니다! 해당내용도 반영해보았습니다.

ohtaeg commented on src/main/java/lotto/model/domain/LottoRun.java:
```diff
@@ -0,0 +1,20 @@
+        return winLotto.compareWithLotto(userLotto);
+    }
+
+    public static double calculateProfit(long buyAmount, long winAmount) {
```
이 로직은 Profit에 있어도 될 것 같은데 윤지님은 어떻게 생각하시나요 :) ?

tlaqk229 replied:
LottoRun 제거하고 해당 로직은 Profit 안에 담았습니다.

ohtaeg reviewed (changes requested):
3단계 구현을 잘해주셨어요  👍 
객체를 잘게 분리해주신게 인상 깊네요 👍 👍 
몇 개의 코멘트들을 드렸는데 다음 미션을 진행하기전에 
같이 고민해보면 좋을 포인트들이 있어서 Request changes 하였습니다!
확인해주신 후 반영해주실 부분이 있다면 반영해주신 후 재요청 해주시면 감사하겠습니다 :)
날씨가 갑지가 추워졌어요 🥶 감기 조심하세요!!

tlaqk229 committed:
fix(PurchaseInfo): 구입 개수 관련 수정

- 기존: 구입금액을 통해 구입개수를 계산, vo에 저장하던 방식
- 수정: 구입금액, 구입개수 두 가지를 하나의 vo에 저장

- vo 생성 시 구입금액을 입력하면 개수를 자동으로 계산하도록 작성
- 구입금액, 구입개수 유효성 체크 추가

tlaqk229 committed:
refactor(LottoRun): 불필요한 호출 개선

- 불필요하게 LottoRun의 메소드를 통해 WinLotto의 메소드를 호출
- LottoRun의 메소드 제거, WinLotto 메소드 직접 호출

tlaqk229 committed:
refactor(Profit): 수익률 계산 로직 -> Profit 객체 안에 구현

tlaqk229 committed:
refactor(package): 정확하지 않은 패키지 명명 수정

tlaqk229 committed:
refactor(LottoResult): dto setter 제거, 생성자 사용하도록 수정

tlaqk229 committed:
refactor(controller): 로또 구입 정보 입력, 세팅 로직 메소드 분리

tlaqk229 committed:
refactor(view): vo 안에 구현된 출력 로직 ResultView로 이동

tlaqk229 committed:
refactor(LottoNumber): 정수 비교 Integer compare 활용

tlaqk229 committed:
refactor(all): 자바 코드 컨벤션 - code reformat, import 최적화

tlaqk229 committed:
docs(학습): 3단계 미션진행하면서 학습한 내용 정리

tlaqk229 commented:
코멘트 달아주신 내용관련하여 고민하면서 찾아본 내용들을 docs안에 정리해보았습니다. 이 부분도 혹시 코멘트해주실 내용있는지 한번 봐주시면 감사하겠습니다!

ohtaeg commented on docs/Domain_VO_study.md:
```diff
@@ -0,0 +1,43 @@
+# 도메인과 VO (feat. DTO)
```
전체적으로 잘 정리해주셨어요 👍 
getter 관련해서 캡슐화와 같이 생각해보시는 것도 추천드립니다 😄

ohtaeg commented on docs/Software_design.md:
```diff
@@ -0,0 +1,37 @@
+# 소프트웨어 설계, 구성
+
+## MVC 패턴
```
👍 👍

ohtaeg commented on src/main/java/lotto/model/domain/PurchaseInfo.java:
```diff
@@ -0,0 +1,74 @@
+public class PurchaseInfo {
+
+    long purchaseAmount;
+    long purchaseCount;
```
이 부분은 같이 고민해보면 좋을 것 같아요!
purchaseCount를 인스턴스 필드로 가지도록 해주신 의도가 있으실까요 :) ?
purchaseCount를 가지고 각 메서드들이 활용하는 측면에서 보면 지속적으로 활용하는 곳이 없는 것 같아 굳이 안가져도 되지 않을까 해서요!

tlaqk229 replied:
purchaseCount는 LottoService의 generateAutoLotto, 그리고 ResultView의 printPurchaseCount 두 곳에서 사용되고 있습니다. 값의 변동이 있지 않은 필드라고 생각했고 필요한 시점마다 계산하지는 않아도 된다고 생각해서 인스턴스 필드로 저장해두도록 하였는데요.  저장해두는 것보다는 필요할 때 계산을 하도록 하는 것이 좋을까요? 일단 그대로 두었습니다!

ohtaeg replied:
이 부분은 윤지님이 생각하시는 방향으로 해주시는게 좋을 것 같아요 :)
개인적으로 내부 필드를 잘 활용해야하는 곳은 내부 필드를 관리하는 PurchaseInfo여야 하는데 내부에서 사용을 하지 않고있어서
필요할 때 계산을 하도록 하는 게 낫지 않을까? 라는 생각으로 드렸던 피드백입니다!
그대로 두셔도 됩니다 👍

ohtaeg commented on src/main/java/lotto/model/domain/PurchaseInfo.java:
```diff
@@ -0,0 +1,74 @@
+
+public class PurchaseInfo {
+
+    long purchaseAmount;
```
접근제어자를 최소화한다면 어떤 장점이 있을까요 :) ?

tlaqk229 replied:
접근제어자 선언하는 걸 빠뜨렸네요ㅜㅜ 수정했습니다!

ohtaeg commented on src/main/java/lotto/model/domain/PurchaseInfo.java:
```diff
@@ -0,0 +1,74 @@
+        checkLottoUnitPrice(); // 개수 유효성 체크
+        checkChange(purchaseAmount); // 잔액 발생 확인
+        //개수 계산
+        this.purchaseCount = calculatePurchaseCount(purchaseAmount);
+        validatePositive(purchaseCount);
```
개인적으로 Validation은 객체 생성을 위한 과정중 하나라고 생각하기에 값을 세팅해주기전에 validation을 먼저 수행해주는 것이 흐름상 괜찮다고 생각하는데 윤지님은 어떻게 생각하시나요 :) ?

tlaqk229 replied:
저도 이부분 작성하면서 고민했던 지점인데 purchaseCount가 계산되어야 purchaseCount의 유효성 체크가 가능한 부분이라 validation을 계산 후에 두게 되었는데요. 코멘트 주셔서 다시 고민해보니 이 부분은 계산하기 전에 purchaseAmount에 대한 유효성 검증을 통해 확인할 수 있는 부분이라 purchaseAmount가 양수인지 확인하는 과정을 앞에 추가하는 방식으로 수정했습니다.

ohtaeg commented on src/main/java/lotto/model/domain/Profit.java:
```diff
@@ -0,0 +1,14 @@
+
+    private double profit;
+
+    public Profit(long buyAmount, long winAmount) {
+        this.profit = Math.floor(winAmount * 1.0 / buyAmount * 100) / 100.0;
+    }
```
잘 옮겨주셨습니다 👍

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -42,7 +42,7 @@ private void setLottoNumber(int lottoNumber) {
     @Override
     public int compareTo(LottoNumber o) {
-        return this.lottoNumber - o.lottoNumber;
+        return Integer.compare(this.lottoNumber, o.lottoNumber);
```
피드백 반영 👍

ohtaeg commented on src/main/java/lotto/model/domain/Lotto.java:
```diff
@@ -64,14 +63,8 @@ public void sortNumbers() {
-        System.out.print(sj);
-        System.out.println(LottoConstants.LOTTO_PRINT_END);
+    public List<LottoNumber> getLotto() {
+        return lotto;
```
Collections.unmodifiablelist()를 활용하여 외부에서 Collection을 변조할 수 없도록 read-only로 리턴하면 좋을 것 같아요 :)

tlaqk229 replied:
이 부분 미처 생각하지 못하였네요! 반영하였습니다! Lottos의 getLottos()도 마찬가지 상황이라고 생각되어 동일하게 수정했습니다.

ohtaeg commented on src/main/java/lotto/model/domain/Lotto.java:
```diff
@@ -64,14 +63,8 @@ public void sortNumbers() {
-            sj.add(lottoNumber.getLottoNumber() + "");
-        }
-        System.out.print(sj);
-        System.out.println(LottoConstants.LOTTO_PRINT_END);
```
피드백 반영 👍

ohtaeg commented on src/main/java/lotto/service/LottoServiceImpl.java:
```diff
@@ -2,40 +2,27 @@
+import lotto.model.domain.WinLotto;
+import lotto.model.domain.WinResult;
 
 public class LottoServiceImpl implements LottoService {
```
수정을 위한 리뷰가 아니오니 참고만 해주시면 감사하겠습니다 :)

별도의 서비스를 만드셔서 로직을 처리하신 점이 인상이 깊습니다 👍 

이번 과정은 사실 객체지향 설계가 목적이기에 MVC 패턴도 좋지만 객체간의 의존 관계를 설정하여
누가 어떤 객체를 생성하고 어떤 객체와 의존하면 좋을지, 객체는 어떤 데이터를 가지고 어떤 행위를 하도록 설계하는 방향이
앞으로 미션을 하시는데 좀 더 도움이 될 거라고 생각하여 조심스럽게 코멘트를 드려봅니다..!

시간이 괜찮으시면 추후에라도 LottoServiceImpl에 구현되어 있는 메서드들을 누가 처리하면 좋을지 고민해보시면 좋을 것 같아요! (입력과 도메인간 중재를 해줘야하는 부분은 어쩔수 없기에 컨트롤러는 있을 수 있다고 생각합니다! )

tlaqk229 replied:
피드백 감사합니다! 고민해보겠습니다!

ohtaeg commented on src/main/java/lotto/view/ResultView.java:
```diff
@@ -25,11 +46,21 @@ public void printLottoResult(LottoResult lottoResult) {
 
     private void printWinResult(WinResult winResult) {
-        winResult.print();
+        Map<WinCriterion, Integer> winResultInfo = winResult.getWinResult();
+        List<WinCriterion> keySet = new ArrayList<>(winResultInfo.keySet());
+        Collections.sort(keySet);
```
뷰가 생각보다 도메인 객체의 내부를 잘 알고있는 것 같아요 :)
getWinResult()를 해주실 때 keySet만 반환하도록 해줘도 괜찮을 것 같은데 윤지님은 어떻게 생각하시나요!? 😄

tlaqk229 replied:
조언해주신 대로 keySet만 반환하는 메소드를 추가하여 활용하였고, 해당 key에 대한 value를 가져오는 것도 필요하여 value를 반환해주는 메소드도 추가하여 활용해보았습니다!

ohtaeg commented on src/test/java/lotto/model/domain/ProfitTest.java:
```diff
@@ -0,0 +1,19 @@
+    @DisplayName("수익률 계산")
+    @Test
+    void 수익률_계산() {
+        long buyAmount = 14000;
+        long winAmount = 5000;
+        double expected = 0.35;
+        Profit profit = new Profit(buyAmount, winAmount);
+        assertThat(profit.getProfit()).isEqualTo(expected);
```
BDD 패턴도 한번 학습해보시면 좋을 것 같아요 😄 

https://www.devkuma.com/docs/bdd/

tlaqk229 replied:
감사합니다! 학습해보겠습니다 😃

ohtaeg reviewed (changes requested):
3단계 피드백 반영을 잘해주셨어요 👍
피드백을 잘 정리해주셨고 의견을 잘 공유해주셨습니다 💯  
추가로 고민해보면 좋을 포인트들이 존재하여 몇 개의 코멘트를 드렸는데요!
남은 미션들을 구현하면서 피드백을 고민하는 것도 좋지만 피드백을 집중해서 고민하면 좋을 것 같아 
Request changes 하였습니다..! 조금만 더 화이팅입니다 🔥 💪

tlaqk229 committed:
fix(PurchaseInfo): 유효성 검증 부분 수정

- 값의 계산 이전에 유효성을 검증하는 방식으로 수정

tlaqk229 committed:
fix(domain): getter로 반환 시 원본 객체가 수정되지 않도록 조치

- 도메인 객체 내의 List 인스턴스 변수를 반환하면 원본을 참조하여
수정이 일어날 수 있음
- 원본과의 참조를 끊고, 반환해준 List 객체 또한 unmodifiable로 처리하여
변조를 막음

tlaqk229 committed:
refactor(view): 도메인 객체 내부 값 접근, 활용 방식 수정

ohtaeg commented on src/main/java/lotto/model/domain/Lottos.java:
```diff
@@ -27,6 +28,6 @@ public MatchCounts compareWinLotto(WinLotto winLotto) {
 
     public List<Lotto> getLottos() {
-        return lottos;
+        return Collections.unmodifiableList(new ArrayList<>(lottos));
```
새 List로 안감싸도 될 것 같아요 :)
```suggestion
        return Collections.unmodifiableList(lottos);
```

ohtaeg commented on src/main/java/lotto/model/domain/WinResult.java:
```diff
@@ -18,7 +20,11 @@ public long calculateWinAmount() {
 
-    public Map<WinCriterion, Integer> getWinResult() {
-        return winResult;
+    public List<WinCriterion> getWinCriterionFromWinResult() {
+        return new ArrayList<>(winResult.keySet());
+    }
```
피드백 반영 👍

ohtaeg commented on src/main/java/lotto/model/domain/PurchaseInfo.java:
```diff
@@ -6,17 +6,23 @@
 
-    long purchaseAmount;
-    long purchaseCount;
+    private long purchaseAmount;
+    private long purchaseCount;
```
👍

ohtaeg commented on src/main/java/lotto/view/ResultView.java:
```diff
@@ -46,12 +46,11 @@ public void printLottoResult(LottoResult lottoResult) {
-            System.out.print(key.getMatchCount() + "개 일치 (" + key.getPrize() + "원)");
-            System.out.println("- " + winResultInfo.get(key) + "개");
+        List<WinCriterion> winCriteria = winResult.getWinCriterionFromWinResult();
+        Collections.sort(winCriteria);
```
WinResult에서 정렬해서 주는 것도 괜찮을 것 같네요 😄

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨습니다 👍 👍 
의견 공유도 적극적으로 해주셔서 감사합니다 :)
다음 미션도 화이팅입니다 🔥
